javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll

```diff
@@ -737,3 +737,106 @@ private module MarsDB {
+ */
+private module Redis {
+  /**
+   * Gets a `Node Redis` client.
+   */
+  private API::Node client() {
+    result = API::moduleImport("redis").getMember("createClient").getReturn()
+    or
+    result = API::moduleImport("redis").getMember("RedisClient").getInstance()
+    or
+    result = client().getMember("duplicate").getReturn()
+    or
+    result = client().getMember("duplicate").getLastParameter().getParameter(1)
+  }
+
+  /**
+   * Gets a (possibly chained) reference to a batch operation object.
+   * These have the same API as a redis client, except the calls are chained, and the sequence is terminated with a `.exec` call.
+   */
+  private API::Node multi() {
+    result = client().getMember(["multi", "batch"]).getReturn()
+    or
+    result = multi().getAMember().getReturn()
+  }
+
+  /**
+   * Gets a `Node Redis` client instance. Either a client created using `createClient()`, or a batch operation object.
+   */
+  private API::Node redis() { result = [client(), multi()] }
+
+  /**
+   * Provides signatures for the query methods from Node Redis.
+   */
+  module QuerySignatures {
+    /**
+     * Holds if `method` interprets parameter `argIndex` as a key, and a later parameter determines a value/field.
+     * Thereby the method is vulnerable if parameter `argIndex` is unexpectedly an array instead of a string, as an attacker can control arguments to Redis that the attacker was not supposed to control.
+     *
+     * Only setters and similar methods are included.
+     * For getter like methods it is not generally possible to gain access "outside" of where you are supposed to have access,
+     * it is at most possible to get a Redis call to return more results than expected (e.g. by adding more members to [`geohash`](https://redis.io/commands/geohash)).
+     */
+    bindingset[argIndex]
+    predicate argumentIsAmbiguousKey(string method, int argIndex) {
+      method =
+        ["set", "publish", "append", "bitfield", "decrby", "getset", "hincrby", "hincrbyfloat",
+            "hset", "hsetnx", "incrby", "incrbyfloat", "linsert", "lpush", "lpushx", "lset",
+            "ltrim", "rename", "renamenx", "rpushx", "setbit", "setex", "smove", "zincrby",
+            "zinterstore", "hdel", "lpush", "pfadd", "rpush", "sadd", "sdiffstore", "srem"] and
+      argIndex = 0
+      or
+      method = ["bitop", "hmset", "mset", "msetnx", "geoadd"] and argIndex >= 0
+    }
+  }
+
+  /**
+   * An expression that is interpreted as a key in a Node Redis call.
+   */
+  class RedisKeyArgument extends NoSQL::Query {
+    RedisKeyArgument() {
+      exists(string method, int argIndex |
+        QuerySignatures::argumentIsAmbiguousKey(method, argIndex)
+      |
+        this = promisify(redis().getMember(method)).getACall().getArgument(argIndex).asExpr()
+      )
+    }
+  }
+
+  /**
+   * Gets a possibly promisified version of `method`.
+   */
+  private API::Node promisify(API::Node method) {
```

Comment:
Entirely reaonable; I've opened an internal issue.

Reply:
> In general it's problematic to map data-flow nodes back to API graph nodes as you lose context. If the node we care about is promisified anywhere in the code, we will get back every promisify in the application. That is, promisify(x) would return both x2,y2 in this case:

Would it? 
I added a test, and that works just fine. 

> Another minor issue is that the predicate contains a tuple for every API node that exists, which isn't a huge problem on its own I guess, but it does seem like an unnecessary cost.

:+1: I changed it. 

> but it would be a great general improvement if we can make API graphs handle promisify transparently. 

Agreed.

Reply:
Ah, we may have been saved by the fact that there can be multiple API nodes for a given access path, so the calls to `util.promisify()` all have separate API nodes.

I still tend to think of API nodes as access paths and the non-uniqueness being an internal detail, but maybe it's a useful detail to embrace.

Anyway, thanks for adding the test, this should be good enough for this PR 👍 .